price-estimator/src/orderbook.rs

```diff
@@ -1,50 +1,98 @@
-        let (account_state, orders) = self
-            .orderbook_reading
-            .get_auction_data(batch_id.into())
-            .await?;
-
-        // TODO: Move this cpu heavy computation out of the async function using spawn_blocking.
-        let pricegraph = Pricegraph::new(
-            orders
-                .iter()
-                .map(|order| order.to_element_with_accounts(&account_state)),
-        );
-
-        Ok(pricegraph)
-    }
-
-    pub async fn get_pricegraph(&self, batch_id: Option<BatchId>) -> Result<Pricegraph> {
+    pub async fn pricegraph(
+        &self,
+        batch_id: Option<BatchId>,
+        pricegraph_type: PricegraphType,
+    ) -> Result<Pricegraph> {
         match batch_id {
-            Some(batch_id) => self.create_pricegraph_at_batch(batch_id).await,
-            None => Ok(self.pricegraph.read().await.clone()),
+            Some(batch_id) => {
+                let mut auction_data = self.auction_data(batch_id).await?;
+                if matches!(pricegraph_type, PricegraphType::WithRoundingBuffer) {
+                    self.apply_rounding_buffer_to_auction_data(&mut auction_data)
+                        .await;
+                }
+                Ok(pricegraph_from_auction_data(&auction_data))
+            }
+            None => Ok(self.cached_pricegraph(pricegraph_type).await),
         }
     }
 
     /// Recreate the pricegraph orderbook.
     pub async fn update(&self) -> Result<()> {
-        let pricegraph = self.create_pricegraph_at_batch(BatchId::now()).await?;
-        *self.pricegraph.write().await = pricegraph;
+        let mut auction_data = self.auction_data(BatchId::now()).await?;
+
+        // TODO: Move this cpu heavy computation out of the async function using spawn_blocking.
+        let pricegraph = pricegraph_from_auction_data(&auction_data);
+        *self.pricegraph_raw.write().await = pricegraph;
+
+        self.apply_rounding_buffer_to_auction_data(&mut auction_data)
+            .await;
+        let pricegraph = pricegraph_from_auction_data(&auction_data);
+        *self.pricegraph_with_rounding_buffer.write().await = pricegraph;
+
         Ok(())
     }
+
+    async fn auction_data(&self, batch_id: BatchId) -> Result<AuctionData> {
+        self.orderbook_reading
+            .get_auction_data(batch_id.into())
+            .await
+    }
+
+    async fn apply_rounding_buffer_to_auction_data(&self, _auction_data: &mut AuctionData) {
+        // TODO
+    }
+
+    async fn cached_pricegraph(&self, pricegraph_type: PricegraphType) -> Pricegraph {
+        match pricegraph_type {
+            PricegraphType::Raw => &self.pricegraph_raw,
+            PricegraphType::WithRoundingBuffer => &self.pricegraph_with_rounding_buffer,
+        }
+        .read()
+        .await
+        .clone()
+    }
+}
+
+type AuctionData = (AccountState, Vec<Order>);
+
+fn pricegraph_from_auction_data(auction_data: &AuctionData) -> Pricegraph {
```

Comment:
Why not use a From/Into trait here? Seems like a pretty standard conversion (we might do a similar thing in the core lib)

Reply:
Here this is not possible because you cannot implement a trait from another crate for a type that is also from another crate.
 We could do this in the core crate but even then to me it doesn't feel appropriate. This feeling is  vague but what I mean is that while it's true that you happen to be able to turn `AuctionData` into `Pricegraph` these types are so unrelated that I don't think about this as a conversion to which the conversion traits apply. It seems more incidental that this works out but it could easily be the case that the pricegraph constructor needed a little more information that isn't part of AuctionData.

Technically any function that takes one type and returns another could be implemented as a conversion trait but semantically it doesn't always fit imo.

Reply:
Oh I forgot that we are actually creating the type AuctionData in this file. Argument above still holds and when we are only doing this in this file then I also prefer not to go through the trait. It is simpler to just have a function. The trait makes sense if you want to provide this convenience for other users of your type.

Reply:
I think the value of using `Into` and `From` is mostly is situations where you want to accept either an AuctionData or Pricegraph by value. This is not the case in our situation since we need to typically use the Pricegraph by reference and construct it before hand.

I also slightly prefer using a function here.